### DISCUSSION TITLE
Is list::copy specific only for some types?

### DISCUSSION THREAD

**Discussion by ales-tsurko** [koto-lang/koto#89] [Q&A]
I've found that `copy` and `deep_copy` of list doesn't work if item is iterator:

```coffee
foo = [(0..10).iter()]
...
» foo.deep_copy().get(0).next()
0
» foo.deep_copy().get(0).next()
1
» foo.deep_copy().get(0).next()
2
```

Also, is there a way to copy an iterator directly?

**Comment by irh** [comment]
Hi @ales-tsurko 👋

Good point, I think that's surprising behaviour and I'd also expect copies of the iterator to be made with `deep_copy`.

My feeling is that generally iterators shouldn't need a special operation to make a copy, they should just be copyable by default.

e.g. 

```coffee
x = (1..100).iter()
y = x

x.next() # 1
y.next() # 2 - I would expect 1 here

x.next() # 3
z = x

x.next() # 4
z.next() # 5 - I would expect 4 here
```

I can't think of a reason to keep the current behaviour of shared state between copies. I'll create an issue to look into it.
